File: src/cs/heap_sort.cs
using System;

public class Program
{
	public static void Main()
	{ 
        int[] seq = GenData(10);
        HeapSort(seq);
        Print(seq);
	}

    private static void HeapSort(int[] seq)
    { 
        for(int i = seq.Length / 2 - 1; i >= 0; i--)
            Sink(seq, i, seq.Length);
        int cnt = seq.Length - 1;
        while(cnt >= 0)
        {
            Swap(seq, 0, cnt);
            Sink(seq, 0, cnt--);
        }
    }

    private static void Sink(int[] seq, int k, int n)
    { 
        while(k * 2 + 1 <= n - 1)
        {
            int j = k * 2 + 1;
            if(j + 1 <= n - 1 && seq[j] < seq[j + 1]) j++;
            if(seq[k] > seq[j]) break;
            Swap(seq, j, k);

            k = j;
        }
    }

    private static int[] GenData(int size)
    {
        int[] seq = new int[size];
        for(int i = 0; i < seq.Length; i++)
            seq[i] = i;

        Random rand = new Random();
        for(int i = 1; i < seq.Length; i++)
        {
            int r = rand.Next(i);
            Swap(seq, i, r);
        }

        return seq; 
    }

    private static void Swap(int[] seq, int p, int q)
    {
        int t = seq[p];
        seq[p] = seq[q];
        seq[q] = t;
    }

    private static void Print(int[] seq)
    {
        for(int i = 0; i < seq.Length; i++)
            Console.Write(seq[i] + " ");
        Console.WriteLine();
    }
}

File: src/py/heap_sort.py
from random import randrange

def gen_data():
    seq = range(10)
    for i in range(1, 10):
        r = randrange(i)
        seq[i], seq[r] = seq[r], seq[i]
    return seq

def heap_sort(seq):
    for i in range((len(seq) - 1)/ 2, -1, -1):
        sink(seq, i, len(seq))

    cnt = len(seq)
    while cnt > 0:
        seq[0], seq[cnt - 1] = seq[cnt - 1], seq[0]
        cnt = cnt - 1
        sink(seq, 0, cnt)

def sink(seq, k, n):
    while (k + 1) * 2 - 1 < n:
        j = (k + 1) * 2 - 1
        if j < n - 1 and seq[j] < seq[j + 1]:
            j = j + 1
        if seq[k] > seq[j]: 
            break
        seq[j], seq[k] = seq[k], seq[j]
        k = j

seq = gen_data()
heap_sort(seq)

print seq
